File: frontend/src/services/databases.ts
/**
 * 数据库监控服务模块
 * 提供数据库列表、详情、性能指标和慢查询的 API 调用
 */
import api from './api';

/** 数据库实例 */
export interface DatabaseItem {
  id: number;
  /** 所属主机 ID */
  host_id: number;
  /** 数据库名称 */
  name: string;
  /** 数据库类型（mysql / postgresql / oracle 等） */
  db_type: string;
  /** 运行状态 */
  status: string;
  created_at: string | null;
  updated_at: string | null;
  /** 最新性能指标快照 */
  latest_metrics?: {
    /** 总连接数 */
    connections_total: number | null;
    /** 活跃连接数 */
    connections_active: number | null;
    /** 数据库大小（MB） */
    database_size_mb: number | null;
    /** 慢查询数量 */
    slow_queries: number | null;
    /** 表数量 */
    tables_count: number | null;
    /** 已提交事务数 */
    transactions_committed: number | null;
    /** 已回滚事务数 */
    transactions_rolled_back: number | null;
    /** 每秒查询数 */
    qps: number | null;
    /** 表空间使用率（%） */
    tablespace_used_pct: number | null;
    /** 记录时间 */
    recorded_at: string | null;
  };
}

/** 慢查询记录 */
export interface SlowQuery {
  /** SQL 标识 */
  sql_id: string;
  /** 平均执行时间（秒） */
  avg_seconds: number;
  /** 执行次数 */
  executions: number;
  /** SQL 语句内容 */
  sql_text: string;
}

/** 慢查询响应 */
export interface SlowQueriesResponse {
  database_id: number;
  slow_queries: SlowQuery[];
}

/** 数据库性能指标 */
export interface DatabaseMetric {
  connections_total: number | null;
  connections_active: number | null;
  database_size_mb: number | null;
  slow_queries: number | null;
  tables_count: number | null;
  transactions_committed: number | null;
  transactions_rolled_back: number | null;
  qps: number | null;
  tablespace_used_pct: number | null;
  recorded_at: string | null;
}

/** 数据库列表响应 */
export interface DatabaseListResponse {
  databases: DatabaseItem[];
  total: number;
}

/** 数据库指标历史响应 */
export interface DatabaseMetricsResponse {
  database_id: number;
  /** 查询时间范围（如 1h / 24h / 7d） */
  period: string;
  metrics: DatabaseMetric[];
}

/** 数据库服务 */
export const databaseService = {
  /** 获取数据库列表 */
  list: (params?: Record<string, unknown>) => api.get<DatabaseListResponse>('/databases', { params }),
  /** 获取单个数据库详情 */
  get: (id: number | string) => api.get<DatabaseItem>(`/databases/${id}`),
  /** 获取数据库性能指标历史（默认最近 1 小时） */
  getMetrics: (id: number | string, period = '1h') => api.get<DatabaseMetricsResponse>(`/databases/${id}/metrics`, { params: { period } }),
  /** 获取慢查询列表 */
  getSlowQueries: (id: number | string) => api.get<SlowQueriesResponse>(`/databases/${id}/slow-queries`),
};

File: backend/app/routers/databases.py
"""F064: Database list, detail, and historical metrics endpoints."""
from datetime import datetime, timezone, timedelta

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select, func, desc
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_user
from app.models.db_metric import MonitoredDatabase, DbMetric

router = APIRouter(prefix="/api/v1/databases", tags=["databases"])


def _parse_period(period: str) -> timedelta:
    """Parse period string like '1h', '24h', '7d' to timedelta."""
    p = period.strip().lower()
    if p.endswith("h"):
        return timedelta(hours=int(p[:-1]))
    if p.endswith("d"):
        return timedelta(days=int(p[:-1]))
    if p.endswith("m"):
        return timedelta(minutes=int(p[:-1]))
    return timedelta(hours=1)


@router.get("")
async def list_databases(
    host_id: int | None = Query(None),
    db: AsyncSession = Depends(get_db),
    _user=Depends(get_current_user),
):
    """List monitored databases with latest metrics."""
    query = select(MonitoredDatabase).order_by(desc(MonitoredDatabase.updated_at))
    if host_id:
        query = query.where(MonitoredDatabase.host_id == host_id)
    result = await db.execute(query)
    databases = result.scalars().all()

    items = []
    for mdb in databases:
        # Get latest metric
        latest = await db.execute(
            select(DbMetric)
            .where(DbMetric.database_id == mdb.id)
            .order_by(desc(DbMetric.recorded_at))
            .limit(1)
        )
        metric = latest.scalar_one_or_none()
        item = {
            "id": mdb.id,
            "host_id": mdb.host_id,
            "name": mdb.name,
            "db_type": mdb.db_type,
            "status": mdb.status,
            "created_at": mdb.created_at.isoformat() if mdb.created_at else None,
            "updated_at": mdb.updated_at.isoformat() if mdb.updated_at else None,
        }
        if metric:
            item["latest_metrics"] = {
                "connections_total": metric.connections_total,
                "connections_active": metric.connections_active,
                "database_size_mb": metric.database_size_mb,
                "slow_queries": metric.slow_queries,
                "tables_count": metric.tables_count,
                "transactions_committed": metric.transactions_committed,
                "transactions_rolled_back": metric.transactions_rolled_back,
                "qps": metric.qps,
                "recorded_at": metric.recorded_at.isoformat() if metric.recorded_at else None,
            }
        items.append(item)

    return {"databases": items, "total": len(items)}


@router.get("/{database_id}")
async def get_database(
    database_id: int,
    db: AsyncSession = Depends(get_db),
    _user=Depends(get_current_user),
):
    """Get database detail."""
    result = await db.execute(select(MonitoredDatabase).where(MonitoredDatabase.id == database_id))
    mdb = result.scalar_one_or_none()
    if not mdb:
        raise HTTPException(status_code=404, detail="Database not found")

    # Latest metric
    latest = await db.execute(
        select(DbMetric).where(DbMetric.database_id == mdb.id).order_by(desc(DbMetric.recorded_at)).limit(1)
    )
    metric = latest.scalar_one_or_none()

    data = {
        "id": mdb.id,
        "host_id": mdb.host_id,
        "name": mdb.name,
        "db_type": mdb.db_type,
        "status": mdb.status,
        "created_at": mdb.created_at.isoformat() if mdb.created_at else None,
        "updated_at": mdb.updated_at.isoformat() if mdb.updated_at else None,
    }
    if metric:
        data["latest_metrics"] = {
            "connections_total": metric.connections_total,
            "connections_active": metric.connections_active,
            "database_size_mb": metric.database_size_mb,
            "slow_queries": metric.slow_queries,
            "tables_count": metric.tables_count,
            "transactions_committed": metric.transactions_committed,
            "transactions_rolled_back": metric.transactions_rolled_back,
            "qps": metric.qps,
            "recorded_at": metric.recorded_at.isoformat() if metric.recorded_at else None,
        }
    return data


@router.get("/{database_id}/metrics")
async def get_database_metrics(
    database_id: int,
    period: str = Query("1h"),
    db: AsyncSession = Depends(get_db),
    _user=Depends(get_current_user),
):
    """Get historical metrics for a database."""
    # Verify database exists
    result = await db.execute(select(MonitoredDatabase).where(MonitoredDatabase.id == database_id))
    if not result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Database not found")

    delta = _parse_period(period)
    since = datetime.now(timezone.utc) - delta

    result = await db.execute(
        select(DbMetric)
        .where(DbMetric.database_id == database_id, DbMetric.recorded_at >= since)
        .order_by(DbMetric.recorded_at)
    )
    metrics = result.scalars().all()

    return {
        "database_id": database_id,
        "period": period,
        "metrics": [
            {
                "connections_total": m.connections_total,
                "connections_active": m.connections_active,
                "database_size_mb": m.database_size_mb,
                "slow_queries": m.slow_queries,
                "tables_count": m.tables_count,
                "transactions_committed": m.transactions_committed,
                "transactions_rolled_back": m.transactions_rolled_back,
                "qps": m.qps,
                "recorded_at": m.recorded_at.isoformat() if m.recorded_at else None,
            }
            for m in metrics
        ],
    }
